File: nreduce/benchmark/quicksort.pl
#!/usr/bin/perl -w

sub genlist
{
  my ($max) = @_;
  my @res = ();

  for (my $i = 0; $i < $max; $i++) {
    my $str = 7*$i;
    my $rev = reverse($str);
    push(@res,$rev);
  }

  return @res;
}

sub quicksort
{
  my @list = @_;

  my $pivot = $list[0];

  my @before = ();
  my @after = ();

  for (my $pos = 1; $pos <= $#list; $pos++) {
    my $cur = $list[$pos];
    if ($cur lt $pivot) {
      push(@before,$cur);
    }
    else {
      push(@after,$cur);
    }
  }

  my @res = ();
  if ($#before >= 0) {
    push(@res,quicksort(@before));
  }
  push(@res,$pivot);
  if ($#after >= 0) {
    push(@res,quicksort(@after));
  }

  return @res;
}

my $n = ($#ARGV < 0) ? 1000 : $ARGV[0];
my @input = genlist($n);
my @sorted = quicksort(@input);
foreach $str (@sorted) {
  print "$str\n";
}

File: nreduce/benchmark/quicksort.py
#!/usr/bin/python

import sys

def genlist(max):
  res = []
  for i in range(0,max):
    rev = str(7*i)[::-1]
    res.append(rev)
  return res

def quicksort(list):
  pivot = list[0]
  before = []
  after = []

  for pos in range(1,len(list)):
    cur = list[pos]
    if cur < pivot:
      before.append(cur)
    else:
      after.append(cur)

  res = []
  if len(before) > 0:
    res += quicksort(before)
  res.append(pivot)
  if len(after) > 0:
    res += quicksort(after)

  return res

n = 1000
if (len(sys.argv) >= 2):
  n = int(sys.argv[1])
input = genlist(n)
sorted = quicksort(input)
for s in sorted:
  print s
